src/main/java/frc/robot/Robot.java

```diff
@@ -21,17 +27,33 @@ public class Robot extends TimedRobot {
   private OperatorInterface oi;
   private CommandFactory commandFactory;
   private Command autoCommand;
+  private RobotContext robotContext;
```

Comment:
@mandrews281 that's a good point, i'm not sure what i am thinking here. i'll review

Reply:
Ok i reviewed, i figured out what i had in mind.
Robot needs robotContext so that it can call periodic, so he needs a reference. RobotContext has access to 'all the things', so that means its periodic can access everything together in one place.

The other decisions are mainly trying to really give everyone only what they need to have, and not more-- as much as possible in construtors.

CommandFactory needed RobotContext to get access to driver preferences, but honestly, you're right, looking at the code, we should really just give it DriverPreferences, since it doenst need  it. CommandFactory does of course need all the subsystems-- but note.  The overall idea here was that CommandFactory would probably end up giving out just about everything in RobotContext eventually-- its kind of a hub.

I was really trying to avoid giving Subsystems to anyone, to prevent them access to each other.

We could clean up the construtors quite a bit by just propagating RobotContext to everyone. I think in practice that would be easier to read. the main reason I didnt was that when everyone has Robotcontext you really dont know what they actually use.

So in the end, for this code, i think i'd recommend one of two things:
(1) give only DriverPreferences to CommandFactory. the overall pattern is to be stingy about what we give out, and prevent giving out things wider than we need

(2) change to giving out RobotContext to all the big players. that's probably ok. The danger is that someone gives out access to RobotContext to comannds, and then the commands turn aroudn and access the subsytsems from there.  That's bad because it will not add the sybsystem as a requirement for the command.

Reply:
makes sense.  I didn't catch the periodic need for RobotContext.